### PR TITLE
Use "contest participants" groups membership to check if a contest is finished (remove groups_attempts.finished_at)

### LINKED ISSUE #296: Handle contest stop correctly
* Get rid of contest finish date in every table
  * For former contests:
    * create a `contest_participants_group_id` group for each contest
    * put the participants (teams or users-not-participating-as-a-team) in this group with a expiration matching their finish date
    * remove their partial access to the contest (the same way it was done before) if they still have
  * All functions around `checkSubmissionRightsForTimeLimitedContest`/`getActiveContestInfoForUser` should use the membership expiration (so `close*Contest` fcts do not make sense anymore)

### PR DESCRIPTION
**Pull request by zenovich**
Fixes #296

### CODE REVIEW

File: app/database/item_store.go [hunk 1]
```diff
@@ -309,119 +297,37 @@ func (s *ItemStore) checkSubmissionRightsForTimeLimitedContest(itemID int64, use
...
 			contest_participations.group_id = groups_ancestors_active.ancestor_group_id AND
 			contest_participations.entered_at IS NOT NULL`).
 		Joins(`
 			LEFT JOIN groups_contest_items ON groups_contest_items.item_id = items.id AND
 				groups_contest_items.group_id = groups_ancestors_active.ancestor_group_id`).
 		Group("items.id").
 		Order("MIN(contest_participations.entered_at) DESC").
-		Having("entered_at IS NOT NULL AND finished_at IS NULL").
-		Limit(1).Scan(&results).Error())
+		PluckFirst("items.id", &itemID).Error()
```
Comment (smadbe) [inline app/database/item_store.go]: I don't get this request, you join with `groups_contest_items` for instance which is not used in the result of the request. groups_attempts are not used as well (or for filtering?).

File: db/migrations/73_create_contest_participants_groups.sql [hunk 2]
```diff
@@ -0,0 +1,92 @@
+-- +migrate Up
+# add new type 'ContestParticipants'
+ALTER TABLE `groups` MODIFY COLUMN `type` enum('Class','Team','Club','Friends','Other','UserSelf','Base','ContestParticipants') NOT NULL;
+
+# create a "contest participants" group for each contest
+INSERT INTO `groups` (`name`, `type`, `team_item_id`)
```
[the remark this answers is not in the document]
Comment (zenovich) [reply db/migrations/73_create_contest_participants_groups.sql]: It is temporary and it gets cleared at line 15

File: app/api/contests/enter.feature [hunk 3]
```diff
@@ -136,6 +136,62 @@ Feature: Enters a contest as a group (user self or team) (contestEnter)
...
+      "data": {
+        "duration": "01:01:01",
+        "entered_at": "3019-10-10T10:10:10Z"
+      }
+    }
+    """
+    And the table "groups_attempts" should be:
+      | group_id | item_id | entered_at          | order |
+      | 31       | 50      | 2019-05-29 11:00:00 | 1     |
+      | 31       | 50      | 3019-10-10 10:10:10 | 2     |
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/contests/enter.feature]: fixed

File: app/api/contests/enter.feature [hunk 4]
```diff
@@ -136,6 +136,62 @@ Feature: Enters a contest as a group (user self or team) (contestEnter)
...
+      | 11                | 41             | 0       | 9999-12-31 23:59:59 |
+      | 11                | 51             | 0       | 9999-12-31 23:59:59 |
+      | 21                | 21             | 1       | 9999-12-31 23:59:59 |
+      | 31                | 31             | 1       | 9999-12-31 23:59:59 |
+      | 41                | 41             | 1       | 9999-12-31 23:59:59 |
+      | 51                | 51             | 1       | 9999-12-31 23:59:59 |
+      | 98                | 98             | 1       | 9999-12-31 23:59:59 |
+      | 99                | 31             | 0       | 3019-10-10 13:13:13 |
+      | 99                | 99             | 1       | 9999-12-31 23:59:59 |
+
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/contests/enter.feature]: added the scenario

File: app/database/item_store.go (hunk 1 above)
Comment (zenovich) [reply app/database/item_store.go]: fixed

File: db/migrations/73_create_contest_participants_groups.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply db/migrations/73_create_contest_participants_groups.sql]: Ah yes sorry, I always forget about this "update x join y" which updates "y".

File: app/api/contests/get_qualification_state.go [hunk 5]
```diff
@@ -72,6 +72,9 @@ type contestGetQualificationStateResponse struct {
 //
 //                      * "Half": same but half of the members (ceil-rounded) of the team;
 //
+//                  * 'not_ready' if the participant has a `groups_attempts` row for the item
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/contests/get_qualification_state.go]: fixed

**Pull request merged**